hudi-common/src/main/java/org/apache/hudi/common/util/ClusteringUtils.java

```diff
@@ -258,26 +258,29 @@ public static Option<HoodieInstant> getOldestInstantToRetainForClustering(
                 ? cleanInstant
                 : HoodieTimeline.getCleanRequestedInstant(cleanInstant.getTimestamp()))
             .getEarliestInstantToRetain();
-        String retainLowerBound;
+        String instantTimeLowerBound;
```

Comment:
```if the timeline of a hoodie table is replace1 commit2 clean3, the earliestInstantToRetain of clean3 is commit2, then replace1 is considered ready for archival no matter when it is completed. But if replace1 is completed after clean3, then the replaced files in replace1 are not cleaned, so it should not be archived. This pr fix such case.```

- In the above case, we should have earliestInstantToRetain to be replace1 if it is not completed.  we should guarantee that earliestInstantToRetain should <= earliest pending instant in terms of timeline.

Reply:
OK, we can add such assurance to `HoodieCleaningPolicy.KEEP_LATEST_BY_HOURS`, so earliestInstantToRetain should <= earliest pending instant in terms of timeline. 

But since `HoodieCleaningPolicy.KEEP_LATEST_FILE_VERSIONS` don't have earliestInstantToRetain, we may still need to use StateTransitionTime in `ClusteringUtils.getOldestInstantToRetainForClustering` when clean policy is KEEP_LATEST_FILE_VERSIONS.

I will make some change to this pr later.